### PR TITLE
[hail] Rewrite VCF INFO Parser to not use htsjdk

### PR DESCRIPTION
**Pull request by chrisvittal**
We should now never call into htsjdk during line parsing, only during header parsing.

### CODE REVIEW

File: hail/src/main/scala/is/hail/io/vcf/LoadVCF.scala [hunk 1]
```diff
@@ -831,7 +1181,7 @@ object LoadVCF {
               val vcfLine = new VCFLine(line, arrayElementsRequired)
               rvb.start(t.physicalType)
               rvb.startStruct()
-              present = vcfLine.parseAddVariant(rvb, rgBc.map(_.value), contigRecoding, skipInvalidLoci)
+              present = vcfLine.parseAddVariant(rvb, rgBc.map(_.value), contigRecoding, hasRSID, skipInvalidLoci)
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/io/vcf/LoadVCF.scala]: this feels like the wrong place for hasRSID - shouldn't it be handled in the same place as the filter/qual?

File: hail/src/main/scala/is/hail/io/vcf/HtsjdkRecordReader.scala [hunk 2]
```diff
@@ -37,57 +37,25 @@ class HtsjdkRecordReader(
...
-          rvb.startArray(vc.getFilters.size())
-          val fi = vc.getFilters.iterator()
-          while (fi.hasNext)
-            rvb.addString(fi.next())
-          rvb.endArray()
-        }
-      }
-    }
-
+    // locus, alleles, rsid added via VCFLine
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/io/vcf/HtsjdkRecordReader.scala]: locus and alleles are the key, this reorg seems wrong.

**Comment by cseed** [comment]
Any performance numbers?

**Comment by chrisvittal** [comment]
If there is an improvement, it is very small, running more tests now, but it looks like the new code is in the noise compared to other factors.

**Comment by chrisvittal** [comment]
More extensive cloud tests are showing speedups in the combiner pipeline compared to master, about 15-20 seconds per partition, but that adds up quickly at scale.

File: hail/src/main/scala/is/hail/io/vcf/LoadVCF.scala [hunk 3]
```diff
@@ -546,6 +646,258 @@ final class VCFLine(val line: String, arrayElementsRequired: Boolean) {
...
+    }
+
+    if (infoType.hasField(key)) {
+      rvb.setFieldIndex(infoType.fieldIdx(key))
+      if (infoFlagFieldNames.contains(key))
+        rvb.addBoolean(true)
+      else
+        parseAddInfoField(rvb, infoType.fieldType(key))
+    }
+    skipInfoField()
```
Comment (jigold) [inline hail/src/main/scala/is/hail/io/vcf/LoadVCF.scala]: `skipInfoField` isn't a good name. I thought it meant skip the info field.

File: hail/src/main/scala/is/hail/io/vcf/LoadVCF.scala [hunk 4]
```diff
@@ -546,6 +646,258 @@ final class VCFLine(val line: String, arrayElementsRequired: Boolean) {
...
+
+    while (!endField()) {
+      nextInfoField()
```
Comment (jigold) [inline hail/src/main/scala/is/hail/io/vcf/LoadVCF.scala]: Can you get rid of the code above and just keep this while loop?

File: hail/src/main/scala/is/hail/io/vcf/LoadVCF.scala (hunk 1 above)
Comment (jigold) [reply hail/src/main/scala/is/hail/io/vcf/LoadVCF.scala]: Is the issue that you have to parse the ID before REF and ALT?

File: hail/src/main/scala/is/hail/io/vcf/LoadVCF.scala [hunk 5]
```diff
@@ -813,6 +1161,7 @@ object LoadVCF {
     arrayElementsRequired: Boolean,
     skipInvalidLoci: Boolean
   ): ContextRDD[RVDContext, RegionValue] = {
+    val hasRSID = t.isInstanceOf[TStruct] && t.asInstanceOf[TStruct].hasField("rsid")
```
Comment (jigold) [inline hail/src/main/scala/is/hail/io/vcf/LoadVCF.scala]: Is this for the prune fields push down?

File: hail/src/main/scala/is/hail/io/vcf/LoadVCF.scala [hunk 6]
```diff
@@ -889,38 +1238,58 @@ object LoadVCF {
     rvb: RegionValueBuilder,
     dropSamples: Boolean = false
   ): Unit = {
-    val vc = c.codec.decode(l.line)
-    reader.readVariantInfo(vc, rvb, c.hasRSID, c.hasQual, c.hasFilters, c.infoSignature, c.infoFlagFieldNames)
+    // QUAL
+    if (c.hasQual) {
+      val qstr = l.parseString()
+      if (qstr == ".")
+        rvb.addDouble(-10.0)
```
Comment (jigold) [inline hail/src/main/scala/is/hail/io/vcf/LoadVCF.scala]: Should this not be just set to missing?

File: hail/src/main/scala/is/hail/io/vcf/HtsjdkRecordReader.scala [hunk 7]
```diff
@@ -33,168 +33,4 @@ class HtsjdkRecordReader(
...
-      case (b: Boolean, _: TBoolean) =>
-        rvb.addBoolean(b)
-      case (s: String, _: TString) =>
-        rvb.addString(s)
-      case (s: String, _: TInt32) =>
-        rvb.addInt(s.toInt)
-      case (s: String, _: TFloat64) =>
-        val d = s match {
-          case "nan" => Double.NaN
-          case "-nan" => Double.NaN
```
Comment (jigold) [inline hail/src/main/scala/is/hail/io/vcf/HtsjdkRecordReader.scala]: I don't think your float parsing handles these correctly.

File: hail/src/main/scala/is/hail/io/vcf/LoadVCF.scala [hunk 8]
```diff
@@ -84,6 +102,15 @@ final class VCFLine(val line: String, arrayElementsRequired: Boolean) {
     }
   }
 
+  def endInfoArrayField(p: Int): Boolean = {
+    if (p == line.length)
+      true
+    else {
+      val c = line(p)
+      c == '\t' || c == ';' || c == ','
```
[the remark this answers is not in the document]
Comment (chrisvittal) [reply hail/src/main/scala/is/hail/io/vcf/LoadVCF.scala]: I copied the name from `endFormatArrayField`, it ends an array element.

File: hail/src/main/scala/is/hail/io/vcf/LoadVCF.scala [hunk 9]
```diff
@@ -93,16 +120,31 @@ final class VCFLine(val line: String, arrayElementsRequired: Boolean) {
     }
   }
 
+  def endFilterArrayField(p: Int): Boolean = endInfoField
```
[the remark this answers is not in the document]
Comment (chrisvittal) [reply hail/src/main/scala/is/hail/io/vcf/LoadVCF.scala]: They're the same, I wanted this name to make sense when parsing FILTERS.

File: hail/src/main/scala/is/hail/io/vcf/LoadVCF.scala [hunk 10]
```diff
@@ -193,13 +248,49 @@ final class VCFLine(val line: String, arrayElementsRequired: Boolean) {
...
+  }
+
   def nextField(): Unit = {
     if (pos == line.length)
       parseError("unexpected end of line")
     assert(line(pos) == '\t')
     pos += 1 // tab
   }
 
+  def nextInfoField(): Unit = {
```
[the remark this answers is not in the document]
Comment (chrisvittal) [reply hail/src/main/scala/is/hail/io/vcf/LoadVCF.scala]: No, the error handling could probably be cleaner, but if the assertion below fails then it's not a valid VCF. We carefully control when the `next` functions are called during parsing so we don't need to ask if there is a next field.

File: hail/src/main/scala/is/hail/io/vcf/LoadVCF.scala [hunk 11]
```diff
@@ -546,6 +646,258 @@ final class VCFLine(val line: String, arrayElementsRequired: Boolean) {
...
+      v = v * 10 + numericValue(line(pos))
+      pos += 1
+    }
+    v * mul
+  }
+
+  def parseAddInfoInt(rvb: RegionValueBuilder) {
+    if (!infoFieldMissing()) {
+      rvb.setPresent()
+      rvb.addInt(parseInfoInt())
```
[the remark this answers is not in the document]
Comment (chrisvittal) [reply hail/src/main/scala/is/hail/io/vcf/LoadVCF.scala]: There is. I set everything to missing in `parseAddInfo` below.

File: hail/src/main/scala/is/hail/io/vcf/LoadVCF.scala [hunk 12]
```diff
@@ -546,6 +646,258 @@ final class VCFLine(val line: String, arrayElementsRequired: Boolean) {
...
+      rvb.setPresent()
+      rvb.addDouble(parseInfoString().toDouble)
+    }
+  }
+
+  def parseIntInInfoArray(): Int = {
+    if (endInfoArrayField())
+      parseError("empty integer")
+    var mul = 1
+    if (line(pos) == '-') {
```
[the remark this answers is not in the document]
Comment (chrisvittal) [reply hail/src/main/scala/is/hail/io/vcf/LoadVCF.scala]: Not quite, the condition is different. There is a larger refactor that I could do to parametrize functions like this on their ending characters.

File: hail/src/main/scala/is/hail/io/vcf/LoadVCF.scala [hunk 13]
```diff
@@ -546,6 +646,258 @@ final class VCFLine(val line: String, arrayElementsRequired: Boolean) {
...
+      pos += 1
+    val end = pos
+    line.substring(start, end)
+  }
+
+  def parseDoubleInInfoArray(): Double = parseStringInInfoArray().toDouble
+
+  def parseIntInfoArrayElement() {
+    if (infoArrayFieldMissing()) {
+      abi.addMissing()
```
[the remark this answers is not in the document]
Comment (chrisvittal) [reply hail/src/main/scala/is/hail/io/vcf/LoadVCF.scala]: This is a `MissingArrayBuilder` our regions need to know the sizes of arrays statically, so we use these as buffers when parsing arrays in a VCF.

File: hail/src/main/scala/is/hail/io/vcf/LoadVCF.scala [hunk 14]
```diff
@@ -546,6 +646,258 @@ final class VCFLine(val line: String, arrayElementsRequired: Boolean) {
...
+      pos += 1  // dot
+    } else
+      abi += parseIntInInfoArray()
+  }
+
+  def parseStringInfoArrayElement() {
+    if (infoArrayFieldMissing()) {
```
[the remark this answers is not in the document]
Comment (chrisvittal) [reply hail/src/main/scala/is/hail/io/vcf/LoadVCF.scala]: Yes, Yes.

[the remark this answers is not in the document]
Comment (chrisvittal) [reply hail/src/main/scala/is/hail/io/vcf/LoadVCF.scala]: That being said, I'm not sure it's best to do so, as again it involves parametrizing on the ending condition.

File: hail/src/main/scala/is/hail/io/vcf/HtsjdkRecordReader.scala (hunk 7 above)
Comment (chrisvittal) [reply hail/src/main/scala/is/hail/io/vcf/HtsjdkRecordReader.scala]: You are correct. Right now it handles it exactly as java would, to be properly VCF 4.3 compliant, we would need to handle, `NaN`, `Inf`, and `-Inf`. Which we don't even handle properly right now, just their lower case versions.

**Comment by chrisvittal** [comment]
@cseed 

```python3
import hail as hl
hl.import_vcf(PATH, reference_genome='GRCh38').write('/tmp/vcfmt', overwrite=True)
```

On my laptop with a warmish filesystem cache takes 1 minute for this code and 1:20 for current master. That's actually pretty good.

File: hail/src/main/scala/is/hail/io/vcf/LoadVCF.scala (hunk 8 above)
[the remark this answers is not in the document]
Comment (jigold) [reply hail/src/main/scala/is/hail/io/vcf/LoadVCF.scala]: That's really confusing.

File: hail/src/main/scala/is/hail/io/vcf/LoadVCF.scala (hunk 14 above)
[the remark this answers is not in the document]
Comment (jigold) [reply hail/src/main/scala/is/hail/io/vcf/LoadVCF.scala]: ok.

File: hail/src/main/scala/is/hail/io/vcf/LoadVCF.scala (hunk 4 above)
Comment (jigold) [reply hail/src/main/scala/is/hail/io/vcf/LoadVCF.scala]: Sorry I meant this:
```
    if (infoType.hasField(key)) {
       rvb.setFieldIndex(infoType.fieldIdx(key))
       if (infoFlagFieldNames.contains(key))
         rvb.addBoolean(true)
       else
         parseAddInfoField(rvb, infoType.fieldType(key))
     }
     skipInfoField()
```

File: hail/src/main/scala/is/hail/io/vcf/LoadVCF.scala (hunk 6 above)
Comment (jigold) [reply hail/src/main/scala/is/hail/io/vcf/LoadVCF.scala]: That's annoying.

File: hail/src/main/scala/is/hail/io/vcf/LoadVCF.scala [hunk 15]
```diff
@@ -1320,3 +1707,15 @@ class VCFsReader(
     }
   }
 }
+
+class BufferedLineIterator(bit: BufferedIterator[String]) extends htsjdk.tribble.readers.LineIterator {
```
Comment (jigold) [inline hail/src/main/scala/is/hail/io/vcf/LoadVCF.scala]: Can you move this to the top of the file?

File: hail/src/main/scala/is/hail/io/vcf/HtsjdkRecordReader.scala (hunk 7 above)
Comment (jigold) [reply hail/src/main/scala/is/hail/io/vcf/HtsjdkRecordReader.scala]: Can you make an issue then?

File: hail/src/main/scala/is/hail/io/vcf/LoadVCF.scala (hunk 12 above)
[the remark this answers is not in the document]
Comment (jigold) [reply hail/src/main/scala/is/hail/io/vcf/LoadVCF.scala]: ok.

File: hail/src/main/scala/is/hail/io/vcf/LoadVCF.scala (hunk 15 above)
Comment (chrisvittal) [reply hail/src/main/scala/is/hail/io/vcf/LoadVCF.scala]: sure

File: hail/src/main/scala/is/hail/io/vcf/LoadVCF.scala (hunk 4 above)
Comment (chrisvittal) [reply hail/src/main/scala/is/hail/io/vcf/LoadVCF.scala]: this is the alternative, thoughts?
```scala
    var key = parseInfoKey()
    if (key == ".") {
      if (endField()) {
        rvb.endStruct()
        return
      } else
        parseError(s"invalid INFO key $key")
    }

    while (!endField()) {
      if (key == ".") {
        parseError(s"invalid INFO key $key")
      }
      if (infoType.hasField(key)) {
        rvb.setFieldIndex(infoType.fieldIdx(key))
        if (infoFlagFieldNames.contains(key))
          rvb.addBoolean(true)
        else
          parseAddInfoField(rvb, infoType.fieldType(key))
      }
      skipInfoField()
      if (!endField()) {
        nextInfoField()
        key = parseInfoKey()
      }
    }
```

File: hail/src/main/scala/is/hail/io/vcf/HtsjdkRecordReader.scala (hunk 7 above)
Comment (chrisvittal) [reply hail/src/main/scala/is/hail/io/vcf/HtsjdkRecordReader.scala]: Futhermore, VCF 4.2 makes no claims about the string representation of floats at all. VCF is garbage.

**Comment by chrisvittal** [comment]
I added some old vs. new tests.

File: hail/python/test/hail/methods/test_impex.py [hunk 16]
```diff
@@ -289,6 +289,22 @@ def test_import_vcfs_subset(self):
         self.assertTrue(vcf2._same(filter1))
         self.assertEqual(len(parts), vcf2.n_partitions())
 
+    def test_vcf_parser_golden_master(self):
+        # the three matrix tables referenced here were generated using the old VCF parser
+        # parser
```
Comment (jigold) [inline hail/python/test/hail/methods/test_impex.py]: delete

File: hail/src/test/resources/ex.vcf [hunk 17]
```diff
@@ -0,0 +1,23 @@
...
+##INFO=<ID=H2,Number=0,Type=Flag,Description="HapMap2 membership">
+##INFO=<ID=HOMSEQ,Number=.,Type=String>
+##FILTER=<ID=q10,Description="Quality below 10">
+##FILTER=<ID=s50,Description="Less than 50% of samples have data">
+##FORMAT=<ID=GT,Number=1,Type=String,Description="Genotype">
+##FORMAT=<ID=GQ,Number=1,Type=Integer,Description="Genotype Quality">
+##FORMAT=<ID=DP,Number=1,Type=Integer,Description="Read Depth">
+##FORMAT=<ID=HQ,Number=2,Type=Integer,Description="Haplotype Quality">
+##FORMAT=<ID=CNL,Number=.,Type=Integer>
+#CHROM	POS	ID	REF	ALT	QUAL	FILTER	INFO	FORMAT	NA00001	NA00002	NA00003
```
Comment (jigold) [inline hail/src/test/resources/ex.vcf]: Does this cover all possible field types? For backwards compatibility add the fields with nans etc. What about adding an example with DB=0? I see arrays of doubles? Do we also need arrays of strings? Can we also get a completely missing INFO field `.`? Just trying to make sure we got all possible code paths.

File: hail/src/main/scala/is/hail/io/vcf/LoadVCF.scala (hunk 4 above)
Comment (jigold) [reply hail/src/main/scala/is/hail/io/vcf/LoadVCF.scala]: This seems better. So we just skip a field silently if it's not in the `infoType`?

File: hail/python/test/hail/methods/test_impex.py (hunk 16 above)
Comment (chrisvittal) [reply hail/python/test/hail/methods/test_impex.py]: ok, I still feel that knowing how these files were generated if this test ever fails will be useful.

**Pull request merged**